tests/unit/s2n_post_handshake_send_test.c

```diff
@@ -0,0 +1,233 @@
+
+    *actual_record_count = cur_seq_num - *last_seq_num;
+    *last_seq_num = cur_seq_num;
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_get_nst_message_size(struct s2n_connection *conn, uint32_t *size)
+{
+    DEFER_CLEANUP(struct s2n_stuffer nst_message = { 0 }, s2n_stuffer_free);
+    RESULT_GUARD_POSIX(s2n_stuffer_growable_alloc(&nst_message, 0));
+    RESULT_GUARD(s2n_tls13_server_nst_write(conn, &nst_message));
+    *size = s2n_stuffer_data_available(&nst_message);
+    EXPECT_TRUE(*size > 50);
+    return S2N_RESULT_OK;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default_tls13"));
+
+    /* These tests require sending tickets.
+     * Setup the config to handle tickets, but don't send any by default.
+     */
+    uint8_t ticket_key_name[16] = "key name";
+    uint8_t ticket_key[] = "key data";
+    uint64_t current_time = 0;
+    EXPECT_SUCCESS(config->wall_clock(config->sys_clock_ctx, &current_time));
+    EXPECT_SUCCESS(s2n_config_set_session_tickets_onoff(config, 1));
+    EXPECT_SUCCESS(s2n_config_add_ticket_crypto_key(config, ticket_key_name, sizeof(ticket_key_name),
+            ticket_key, sizeof(ticket_key), current_time / ONE_SEC_IN_NANOS));
+    config->initial_tickets_to_send = 0;
+
+    const uint32_t fragment_sizes[] = {
+        1,
+        2,
+        TLS_HANDSHAKE_HEADER_LENGTH,
+        TLS_HANDSHAKE_HEADER_LENGTH + 1,
+        50,
+        S2N_DEFAULT_FRAGMENT_LENGTH,
+        S2N_TLS_MAXIMUM_FRAGMENT_LENGTH,
+    };
+
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    const uint8_t send_data[1] = { 'k' };
+    const uint8_t tickets_to_send = 3;
+
+    /* Test: send fragmented post-handshake message (NewSessionTicket) */
+    {
+        DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+        EXPECT_SUCCESS(s2n_connection_set_all_protocol_versions(server_conn, S2N_TLS13));
+        EXPECT_OK(s2n_connection_set_secrets(server_conn));
+
+        DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+        EXPECT_SUCCESS(s2n_connection_set_all_protocol_versions(client_conn, S2N_TLS13));
+        EXPECT_OK(s2n_connection_set_secrets(client_conn));
+
+        DEFER_CLEANUP(struct s2n_test_io_stuffer_pair io_pair = { 0 }, s2n_io_stuffer_pair_free);
+        EXPECT_OK(s2n_io_stuffer_pair_init(&io_pair));
+        EXPECT_OK(s2n_connections_set_io_stuffer_pair(client_conn, server_conn, &io_pair));
+
+        /* Calculate the size of the NewSessionTicket message */
+        uint32_t nst_size = 0;
+        EXPECT_OK(s2n_get_nst_message_size(server_conn, &nst_size));
```

Comment:
Do you think we should have a test for larger NST sizes? I don't remember how large ours are currently, but if they're too small we might not be actually testing the fragmentation logic.

Reply:
I can add early data information to make the NST larger, but I'm not sure that's particularly useful.

We're definitely testing the fragmentation logic. NSTs are ~140 bytes. I'm testing with very small fragment sizes (1, 2, 4, 5, etc). I also had "50" in there, and asserted that the NSTs were >50 bytes, because I wanted a larger fragment size that would still trigger fragmenting. To make that clear I've named the "50" value rather than just hardcoding it.